File: back/app/api/routers/quote_drafts.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from uuid import UUID
from typing import List

from app.api.deps import get_db, get_current_user
from app.models import User, QuoteDraft
from app.schemas.quote_draft import (
    QuoteDraftCreate,
    QuoteDraftUpdate,
    QuoteDraftResponse,
    QuoteDraftListResponse,
)

router = APIRouter(prefix="/quote-drafts", tags=["Quote Drafts"])


@router.get("", response_model=QuoteDraftListResponse)
async def list_quote_drafts(
    folder_id: UUID = Query(..., description="ID du dossier"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Liste les brouillons de devis d'un dossier.
    Filtré automatiquement par tenant.
    """
    # Requête pour récupérer les brouillons avec pagination
    query = select(QuoteDraft).where(
        QuoteDraft.folder_id == folder_id,
        QuoteDraft.tenant_id == current_user.tenant_id
    ).order_by(QuoteDraft.updated_at.desc())
    
    # Compte total
    count_query = select(func.count()).select_from(QuoteDraft).where(
        QuoteDraft.folder_id == folder_id,
        QuoteDraft.tenant_id == current_user.tenant_id
    )
    
    result = await db.execute(query)
    count_result = await db.execute(count_query)
    
    drafts = result.scalars().all()
    total = count_result.scalar_one()
    
    return QuoteDraftListResponse(drafts=drafts, total=total)


@router.post("", response_model=QuoteDraftResponse, status_code=status.HTTP_201_CREATED)
async def create_quote_draft(
    draft_in: QuoteDraftCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Crée un nouveau brouillon de devis.
    """
    # Convertir les lignes en dict pour JSONB
    lines_dict = [line.model_dump() for line in draft_in.lines]
    
    draft = QuoteDraft(
        name=draft_in.name,
        folder_id=draft_in.folder_id,
        module_code=draft_in.module_code,
        product_ids=draft_in.product_ids,
        lines=lines_dict,
        total_ht=draft_in.total_ht,
        total_ttc=draft_in.total_ttc,
        rac_ttc=draft_in.rac_ttc,
        cee_prime=draft_in.cee_prime,
        margin_ht=draft_in.margin_ht,
        margin_percent=draft_in.margin_percent,
        strategy_used=draft_in.strategy_used,
        warnings=draft_in.warnings,
        tenant_id=current_user.tenant_id,
    )
    
    db.add(draft)
    await db.commit()
    await db.refresh(draft)
    return draft


@router.get("/{draft_id}", response_model=QuoteDraftResponse)
async def get_quote_draft(
    draft_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Récupère un brouillon de devis spécifique.
    Vérifie l'appartenance au tenant.
    """
    query = select(QuoteDraft).where(
        QuoteDraft.id == draft_id,
        QuoteDraft.tenant_id == current_user.tenant_id
    )
    result = await db.execute(query)
    draft = result.scalar_one_or_none()

    if not draft:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Brouillon de devis non trouvé."
        )

    return draft


@router.put("/{draft_id}", response_model=QuoteDraftResponse)
async def update_quote_draft(
    draft_id: UUID,
    draft_in: QuoteDraftUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Met à jour un brouillon de devis.
    """
    query = select(QuoteDraft).where(
        QuoteDraft.id == draft_id,
        QuoteDraft.tenant_id == current_user.tenant_id
    )
    result = await db.execute(query)
    draft = result.scalar_one_or_none()

    if not draft:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Brouillon de devis non trouvé."
        )

    update_data = draft_in.model_dump(exclude_unset=True)
    
    # Convertir les lignes en dict si présentes
    if "lines" in update_data and update_data["lines"] is not None:
        update_data["lines"] = [line.model_dump() for line in draft_in.lines]
    
    for field, value in update_data.items():
        setattr(draft, field, value)

    await db.commit()
    await db.refresh(draft)
    return draft


@router.delete("/{draft_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_quote_draft(
    draft_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Supprime un brouillon de devis.
    """
    query = select(QuoteDraft).where(
        QuoteDraft.id == draft_id,
        QuoteDraft.tenant_id == current_user.tenant_id
    )
    result = await db.execute(query)
    draft = result.scalar_one_or_none()

    if not draft:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Brouillon de devis non trouvé."
        )

    await db.delete(draft)
    await db.commit()
    return None

File: front/src/lib/api/quote-drafts.ts
import { api } from "@/lib/api";

// ============================================================================
// Types
// ============================================================================

export interface QuoteLine {
  product_id: string | null;
  title: string;
  description: string;
  quantity: number;
  unit_price_ht: number;
  tva_rate: number;
  total_ht: number;
  total_ttc: number;
  is_editable: boolean;
}

export interface QuoteDraft {
  id: string;
  name: string;
  folder_id: string;
  module_code: string;
  product_ids: string[];
  lines: QuoteLine[];
  total_ht: number;
  total_ttc: number;
  rac_ttc: number;
  cee_prime: number;
  margin_ht: number;
  margin_percent: number;
  strategy_used: string;
  warnings: string[];
  tenant_id: string;
  created_at: string;
  updated_at: string;
}

export interface QuoteDraftCreate {
  name: string;
  folder_id: string;
  module_code: string;
  product_ids: string[];
  lines: QuoteLine[];
  total_ht: number;
  total_ttc: number;
  rac_ttc: number;
  cee_prime: number;
  margin_ht: number;
  margin_percent: number;
  strategy_used: string;
  warnings: string[];
}

export interface QuoteDraftUpdate {
  name?: string;
  product_ids?: string[];
  lines?: QuoteLine[];
  total_ht?: number;
  total_ttc?: number;
  rac_ttc?: number;
  cee_prime?: number;
  margin_ht?: number;
  margin_percent?: number;
  strategy_used?: string;
  warnings?: string[];
}

export interface QuoteDraftListResponse {
  drafts: QuoteDraft[];
  total: number;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Liste les brouillons de devis d'un dossier
 */
export async function listQuoteDrafts(folderId: string): Promise<QuoteDraftListResponse> {
  const response = await api.get(`quote-drafts?folder_id=${folderId}`);
  return response.json();
}

/**
 * Récupère un brouillon de devis spécifique
 */
export async function getQuoteDraft(draftId: string): Promise<QuoteDraft> {
  const response = await api.get(`quote-drafts/${draftId}`);
  return response.json();
}

/**
 * Crée un nouveau brouillon de devis
 */
export async function createQuoteDraft(draft: QuoteDraftCreate): Promise<QuoteDraft> {
  const response = await api.post("quote-drafts", draft as any);
  return response.json();
}

/**
 * Met à jour un brouillon de devis
 */
export async function updateQuoteDraft(
  draftId: string,
  draft: QuoteDraftUpdate
): Promise<QuoteDraft> {
  const response = await api.put(`quote-drafts/${draftId}`, draft as any);
  return response.json();
}

/**
 * Supprime un brouillon de devis
 */
export async function deleteQuoteDraft(draftId: string): Promise<void> {
  await api.delete(`quote-drafts/${draftId}`);
}

/**
 * Génère automatiquement un nom de brouillon basé sur la date
 */
export function generateDraftName(): string {
  const now = new Date();
  const date = now.toLocaleDateString("fr-FR", {
    day: "2-digit",
    month: "2-digit",
    year: "numeric",
  });
  const time = now.toLocaleTimeString("fr-FR", {
    hour: "2-digit",
    minute: "2-digit",
  });
  return `Brouillon du ${date} ${time}`;
}
